runtime-sdk/src/storage/confidential.rs

```diff
@@ -0,0 +1,252 @@
+use std::convert::TryInto as _;
+
+use anyhow::{anyhow, Error};
+
+use crate::{
+    context::Context,
+    core::{
+        common::crypto::{
+            hash::Hash,
+            mrae::deoxysii::{self, NONCE_SIZE},
+        },
+        storage::mkvs,
+    },
+    keymanager::{KeyManagerClientWithContext, KeyPair, KeyPairId},
+    storage::Store,
+};
+
+type Nonce = [u8; NONCE_SIZE];
+
+/// A key-value store that encrypts all content with DeoxysII.
+pub struct ConfidentialStore<S: Store> {
+    inner: S,
+    deoxys: deoxysii::DeoxysII,
+}
+
+impl<S: Store> ConfidentialStore<S> {
+    /// Create a new confidential store with the given keypair.
+    pub fn new_with_key_pair(inner: S, keypair: KeyPair) -> Self {
+        let mut actual_key = [0u8; deoxysii::KEY_SIZE];
+        actual_key.copy_from_slice(&keypair.state_key.0[..deoxysii::KEY_SIZE]);
+        ConfidentialStore {
+            inner,
+            deoxys: deoxysii::DeoxysII::new(&actual_key),
+        }
+    }
+
+    pub fn new_with_id_from_client(
+        kmgr: &KeyManagerClientWithContext<'_>,
+        inner: S,
+        kid: KeyPairId,
+    ) -> Result<Self, Error> {
+        let keypair = kmgr.get_or_create_keys(kid)?;
+        Ok(Self::new_with_key_pair(inner, keypair))
+    }
+
+    /// Create a new confidential store with the given keymanager key id.
+    pub fn new_with_id<C: Context>(ctx: &C, inner: S, kid: KeyPairId) -> Result<Self, Error> {
+        let kmgr = ctx
+            .key_manager()
+            .ok_or_else(|| anyhow!("confidential transactions not available"))?;
+        Self::new_with_id_from_client(kmgr, inner, kid)
+    }
+
+    fn pack_key(&self, enc_key: &[u8], nonce: &Nonce) -> Vec<u8> {
+        let mut ret = Vec::with_capacity(nonce.len() + enc_key.len());
+        ret.extend_from_slice(nonce);
+        ret.extend_from_slice(enc_key);
+        ret
+    }
+
+    fn make_key(&self, plain_key: &[u8]) -> (Nonce, Vec<u8>) {
+        let mut nonce = [0u8; NONCE_SIZE];
+        let plain_hash = Hash::digest_bytes(plain_key);
+        nonce.copy_from_slice(plain_hash.truncated(NONCE_SIZE));
+        let enc_key = self.deoxys.seal(&nonce, plain_key.to_vec(), nonce.to_vec());
```

Comment:
Why is nonce being used as AD here (and elsewhere)? You should probably use `vec![]` for additional data for now.